akka-stream/src/main/scala/akka/stream/impl/TraversalBuilder.scala

```diff
@@ -0,0 +1,985 @@
+        // TODO Optimize Map access
+        pendingBuilders = pendingBuilders.updated(builderKey, result),
+        // pendingBuilders = pendingBuilders - builderKey,
+        unwiredOuts = unwiredOuts - 1
+      )
+    } else {
+      // Update structures with result
+      copy(
+        inBaseOffsetForOut = inBaseOffsetForOut - out,
+        unwiredOuts = unwiredOuts - 1,
+        pendingBuilders = pendingBuilders.updated(builderKey, result)
+      )
+    }
+
+    // If we have no more unconnected outputs, we can finally build the Traversal and shed most of the auxiliary data.
+    wired.completeIfPossible
+  }
+
+  // Requires that a remapped Shape's ports contain the same ID as their target ports!
+  def add[A, B, C](submodule: TraversalBuilder, shape: Shape, combineMat: (A, B) ⇒ C): TraversalBuilder = {
+    val builderKey = new BuilderKey
+
+    val newBuildSteps =
+      if (combineMat == Keep.left) {
+        AppendTraversal(Pop) ::
+          builderKey ::
+          reverseBuildSteps
+      } else if (combineMat == Keep.right) {
+        builderKey ::
+          AppendTraversal(Pop) ::
+          reverseBuildSteps
+      } else if (combineMat == Keep.none) {
+        AppendTraversal(PushNotUsed) ::
+          AppendTraversal(Pop) ::
+          AppendTraversal(Pop) ::
+          builderKey ::
+          reverseBuildSteps
+      } else {
+        AppendTraversal(Compose(combineMat.asInstanceOf[(Any, Any) ⇒ Any])) ::
+          builderKey ::
+          reverseBuildSteps
+      }
+
+    val added = if (submodule.isTraversalComplete) {
+      // We only need to keep track of the offsets of unwired inputs. Outputs have all been wired
+      // (isTraversalComplete = true).
+
+      var newInOffsets = inOffsets
+      val inIterator = shape.inlets.iterator
+      while (inIterator.hasNext) {
+        val in = inIterator.next()
+        // Calculate offset in the current scope. This is the our first unused input slot plus
+        // the relative offset of the input port in the submodule.
+        // TODO Optimize Map access
+        newInOffsets = newInOffsets.updated(in, inSlots + submodule.offsetOf(in.mappedTo))
+      }
+
+      copy(
+        reverseBuildSteps = newBuildSteps,
+        inSlots = inSlots + submodule.inSlots,
+        pendingBuilders = pendingBuilders.updated(builderKey, submodule),
+        inOffsets = newInOffsets
+      )
+    } else {
+      // Added module have unwired outputs.
+
+      var newInOffsets = inOffsets
+      var newBaseOffsetsForOut = inBaseOffsetForOut
+      var newOutOwners = outOwners
+
+      // See the other if case for explanation of this
+      val inIterator = shape.inlets.iterator
+      while (inIterator.hasNext) {
+        val in = inIterator.next()
+        // Calculate offset in the current scope
+        // TODO Optimize Map access
+        newInOffsets = newInOffsets.updated(in, inSlots + submodule.offsetOf(in.mappedTo))
+      }
+
+      val outIterator = shape.outlets.iterator
```

Comment:
if there are cases where inlets or outlets can be empty it might be worth checking `nonEmpty` before creating the iterator, note that even `Nil.iterator` creates a new instance